src/MarlinTrkUtils.cc

```diff
@@ -821,19 +823,48 @@ namespace MarlinTrk {
     
     int detElementID = 0;
     
-    return_error = marlintrk->propagateToLayer(encoder.lowWord(), trkhit, *trkStateCalo, chi2, ndf, detElementID, IMarlinTrack::modeForward ) ;
+    TrackStateImpl tsBarrel;
+    TrackStateImpl tsEndcap;
     
-    if (return_error == IMarlinTrack::no_intersection ) { // try forward or backward
-
-      encoder[lcio::LCTrackerCellID::subdet()] = ecal_endcap_face_ID ;
-
-      if (tanL_is_positive) {
+    // propagate to the barrel layer
+    return_error_barrel = marlintrk->propagateToLayer(encoder.lowWord(), trkhit, tsBarrel, chi2, ndf, detElementID, IMarlinTrack::modeForward ) ;
+    
+    // propagate to the endcap layer
+    encoder[lcio::LCTrackerCellID::subdet()] = ecal_endcap_face_ID ;
+    if (tanL_is_positive) {
         encoder[lcio::LCTrackerCellID::side()] = lcio::ILDDetID::fwd;
-      }
-      else{
+    }
+    else{
         encoder[lcio::LCTrackerCellID::side()] = lcio::ILDDetID::bwd;
-      }
-      return_error = marlintrk->propagateToLayer(encoder.lowWord(), trkhit, *trkStateCalo, chi2, ndf, detElementID, IMarlinTrack::modeForward ) ;
+    }
+    return_error_endcap = marlintrk->propagateToLayer(encoder.lowWord(), trkhit, tsEndcap, chi2, ndf, detElementID, IMarlinTrack::modeForward ) ;
+
+    // check which is the right intersection / closer to the trkhit
+    if ( return_error_barrel == IMarlinTrack::no_intersection ){
+        //if barrel fails just return ts at the Endcap if exists
+        return_error = marlintrk->propagateToLayer(encoder.lowWord(), trkhit, *trkStateCalo, chi2, ndf, detElementID, IMarlinTrack::modeForward );
```

Comment:
I think you could save one extrapolation/propagation here, since you have just done it, and you could just use the results from that, right?
```suggestion
        return_error = return_error_endcap;
        *trkStateCalo = tsEndcap;
```